### ISSUE TITLE
Cronjobs increasing CPU usage and slow queries

### ISSUE THREAD

**Issue by Zyles** [magento/magento2#26507]
Cronjobs stuck in pending state. CPU usage increasing. Mysql slow queries increasing.

![image](https://user-images.githubusercontent.com/1741593/72984536-39d01f00-3de4-11ea-838b-282e03f6fa54.png)

Time is 13:28.

Logs filled with:

```
use magento;
SET timestamp=1579781526;
SELECT GET_LOCK('|CRON_GROUP_default', '5');
# Time: 200123 13:13:06
# User@Host: magento[magento] @ localhost []
# Thread_id: 20330  Schema: magento  QC_hit: No
# Query_time: 2.951003  Lock_time: 0.000000  Rows_sent: 1  Rows_examined: 0
# Rows_affected: 0
SET timestamp=1579781586;
```

Load goes down after disabling cronjob in crontab:

![image](https://user-images.githubusercontent.com/1741593/72985132-941daf80-3de5-11ea-974c-ddf6b6c14ed0.png)

### Preconditions (*)

Magento 2.3.3 & 2.4-develop
PHP-FPM 7.2.24

### Steps to reproduce (*)

1. Install Magento 
2. Setup cron jobs
3. Run store for a week
4. Break store

### Expected result (*)

Normal working store.

### Actual result (*)

1. Increased CPU usage
2. Increased cronjobs count
3. Increased slow queries
4. Slow store

**Comment by gwharton** [comment]
Do ypu have any jobs stuck "running"

**Comment by devchris79** [comment]
We see some cron jobs stuck in the running status:
![cron](https://user-images.githubusercontent.com/43850303/72993029-9afbf080-3dec-11ea-94aa-e1b1bd66594c.jpg)

That's with a Magento 2.3.3 installation.

**Comment by gwharton** [comment]
If you manually delete the rows that are stuck running, the pending jobs should be cleaned back to normal levels over the next few cron runs.

**Comment by gwharton** [comment]
But you need to look at why they do not complete. It would be better if magento had some timeout to terminate stuck running jobs.

**Comment by devchris79** [comment]
I have deleted them out before, but they returned (as I suppose you would expect). I will try again and try to understand the sticking.

I agree with the timeout to terminate stuck routines, that would be great. I also don't understand why we have routines stuck in the "running" state with no execution time stamp, surely that should be the first thing done....

**Comment by Zyles** [comment]
Didn't check "running". Truncated table as it took up too much server resources.

Clearly a bug.

One should not routinely have to empty the table manually.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Hotel 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30804` were created

**Issue Available**: @engcom-Hotel, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by devchris79** [comment]
I have tweaked my cron job setup slightly and so far there hasn't been any sticking in the running state. I am wondering if its possible that two or more instances of the same cron job are being run at the same time and if that's causing lockup.

**Comment by anil90** [comment]
hi @devchris79 we are also facing same issue with Magento 2.3.2 Could you please let us know your changes. Thank you

**Comment by devchris79** [comment]
Hi @anil90 we are still seeing the issues. I think the timing is a red herring.

**Comment by devchris79** [comment]
The issue #26809 contains a post that highlights the extra ram usage in the cron processes from Magento 2.3.3.

Our site currently has no jobs stuck in the running state for over a day, and I have only made two changes:

1, Truncated the cron log as it was getting very large
2, We were seeing errors in the system log relating to aqmp not being configured. We don't use RabbitMQ so I just added the code below to env.php:

```
,
'cron_consumers_runner' => [
        'cron_run' => false,
        'max_messages' => 1000,
        'consumers' => [
            'async.operations.all'
        ]
    ] 
```

**Comment by roni-sooryen** [comment]
@devchris79  Please keep me up to date with the status of your cron job processes. You can use the https://github.com/kiwicommerce/magento2-cron-scheduler to kill your cron job after the cron job runs for 3 hours.

**Comment by devchris79** [comment]
@roni-sooryen Happy to report that we still have nothing stuck :-)

**Comment by roni-sooryen** [comment]
Seems like this is the same issue here https://github.com/magento/magento2/issues/22438

**Comment by Zyles** [comment]
I've had more success using separate intervals on cronjob, so they do not lock into each other. This is on a medium store.

On a large store the problem should be even more severe since the cronjobs run for longer.

Example:

```
* * * * * /usr/bin/php /home/http/magento/bin/magento cron:run 2>&1 | grep -v "Ran jobs by schedule" >> /home/http/magento/var/log/magento.cron.log
10 * * * * /usr/bin/php /home/http/magento/update/cron.php >> /home/http/magento/var/log/update.cron.log
15 * * * * /usr/bin/php /home/http/magento/bin/magento setup:cron:run >> /home/http/magento/var/log/setup.cron.log

```

Also turning off separate processes for each cronjob in backend:

![image](https://user-images.githubusercontent.com/1741593/75478643-9c49ab80-599e-11ea-93a4-92be23a226eb.png)


This is by no means a fix, rather a workaround and works for us right now.

**Comment by ilnytskyi** [comment]
it's possible to install each cron group separately with lock as I described in this comment
https://github.com/magento/magento2/issues/25987#issuecomment-576034263

**Comment by driskell** [comment]
#28007 can help with high CPU as it removes some deadlocks that prevent cron schedule table from being cleaned. It can also fix things stuck in running status.

**Comment by magento-engcom-team** [comment]
Hi @Zyles. Thank you for your report. 
The issue has been fixed in magento/magento2#28007 by @driskell in 2.4-develop branch
Related commit(s):
 - [aa12503923ac994873724ecaa759e1eb14b318dc](https://github.com/magento/magento2/commit/aa12503923ac994873724ecaa759e1eb14b318dc)
 - [b0b72b0f8486ca119015e84b7786dfab13d2e475](https://github.com/magento/magento2/commit/b0b72b0f8486ca119015e84b7786dfab13d2e475)
 - [315bd3af8c5ed3dd711abd40db9a163b0c953855](https://github.com/magento/magento2/commit/315bd3af8c5ed3dd711abd40db9a163b0c953855)
 - [96029b55f7cd00b58967637aedac67ddd1767bad](https://github.com/magento/magento2/commit/96029b55f7cd00b58967637aedac67ddd1767bad)
 - [7df183b6c6bc3f57f0c3d29d279c3c6d631cf455](https://github.com/magento/magento2/commit/7df183b6c6bc3f57f0c3d29d279c3c6d631cf455)
 - [68d679b7bd7d371c5266d4803f03f204893f7af2](https://github.com/magento/magento2/commit/68d679b7bd7d371c5266d4803f03f204893f7af2)
 - [a90161907aedae371cb307759b685cc254efe2d8](https://github.com/magento/magento2/commit/a90161907aedae371cb307759b685cc254efe2d8)
 - [dda5c72c9303f9476ce7041743c87079cada87f6](https://github.com/magento/magento2/commit/dda5c72c9303f9476ce7041743c87079cada87f6)
 - [4d4fa685c0b507b510ad899f5afe94bd6c836115](https://github.com/magento/magento2/commit/4d4fa685c0b507b510ad899f5afe94bd6c836115)
 - [e95d2fefefb62150e566f000a4a17f58f6716af5](https://github.com/magento/magento2/commit/e95d2fefefb62150e566f000a4a17f58f6716af5)
 - [916bc93138f7a81f65c2e8b5069eeb0f35fceb43](https://github.com/magento/magento2/commit/916bc93138f7a81f65c2e8b5069eeb0f35fceb43)
 - [edd77531f8847f7b13dbaca9aefa315ba4aa5e57](https://github.com/magento/magento2/commit/edd77531f8847f7b13dbaca9aefa315ba4aa5e57)
 - [e70027f838f766b816bd1f7063514c1a9caf7a20](https://github.com/magento/magento2/commit/e70027f838f766b816bd1f7063514c1a9caf7a20)
 - [92dc76438063e68f7523d5ab036a57d7faf7bc03](https://github.com/magento/magento2/commit/92dc76438063e68f7523d5ab036a57d7faf7bc03)
 - [d1b22ff825052d836b56f8ce33760d42b2241980](https://github.com/magento/magento2/commit/d1b22ff825052d836b56f8ce33760d42b2241980)
 - [a76482c4614a7660ab2d324ca4c086f66906c701](https://github.com/magento/magento2/commit/a76482c4614a7660ab2d324ca4c086f66906c701)

The fix will be available with the upcoming 2.4.3 release.

**Comment by Adel-Magebinary** [comment]
Still happens in 2.4.2
